x/finality/keeper/power_dist_change.go

```diff
@@ -396,23 +301,136 @@ func (k Keeper) ProcessAllPowerDistUpdateEvents(
+			if typedEvent.BtcDelStateUpdate.NewState == types.BTCDelegationStatus_EXPIRED {
+				// Defer EXPIRED events for later processing
+				state.ExpiredEvents = append(state.ExpiredEvents, typedEvent)
+			} else {
+				// Process ACTIVE/UNBONDED events immediately
+				k.processBtcDelUpdateImmediate(ctx, state, typedEvent)
+			}
+		default:
+			// Process all FP events immediately
+			k.processFPEventImmediate(sdkCtx, state, event)
+		}
+	}
+}
+
+// processBtcDelUpdateImmediate processes a BTC delegation update event immediately.
+// It handles the BTC delegation state update by checking the new state and
+// updating the processing state accordingly.
+func (k Keeper) processBtcDelUpdateImmediate(ctx context.Context, state *ftypes.ProcessingState, event *types.EventPowerDistUpdate_BtcDelStateUpdate) {
+	delEvent := event.BtcDelStateUpdate
+	delStkTxHash := delEvent.StakingTxHash
+
+	btcDel, err := k.BTCStakingKeeper.GetBTCDelegation(ctx, delStkTxHash)
+	if err != nil {
+		panic(err) // only programming error
+	}
+	delParams := k.BTCStakingKeeper.GetParamsByVersion(ctx, btcDel.ParamsVersion)
+
+	switch delEvent.NewState {
+	case types.BTCDelegationStatus_ACTIVE:
+		k.processPowerDistUpdateEventActive(ctx, state, btcDel)
+	case types.BTCDelegationStatus_UNBONDED:
+		// In case of delegation transtioning from phase-1 it is possible that
+		// somebody unbonds before receiving the required covenant signatures.
+		hasQuorum, err := k.BTCStakingKeeper.BtcDelHasCovenantQuorums(ctx, btcDel, delParams.CovenantQuorum)
+		if err != nil {
+			panic(err)
+		}
+		if hasQuorum {
+			// add the unbonded BTC delegation to the map
+			k.processPowerDistUpdateEventUnbond(ctx, state, btcDel)
+		}
+	}
+}
+
+func (k Keeper) processExpiredEvents(ctx context.Context, sdkCtx sdk.Context, state *ftypes.ProcessingState) {
+	for _, event := range state.ExpiredEvents {
+		delEvent := event.BtcDelStateUpdate
+		delStkTxHash := delEvent.StakingTxHash
+
+		btcDel, err := k.BTCStakingKeeper.GetBTCDelegation(ctx, delStkTxHash)
+		if err != nil {
+			panic(err) // only programming error
+		}
+
+		types.EmitExpiredDelegationEvent(sdkCtx, delStkTxHash)
+		if btcDel.IsUnbondedEarly() {
+			continue
+		}
+		delParams := k.BTCStakingKeeper.GetParamsByVersion(ctx, btcDel.ParamsVersion)
+
+		// We process expired event if:
+		// - it hasn't unbonded early
+		// - it has all required covenant signatures
+		hasQuorum, err := k.BTCStakingKeeper.BtcDelHasCovenantQuorums(ctx, btcDel, delParams.CovenantQuorum)
+		if err != nil {
+			panic(err)
+		}
+		if hasQuorum {
+			// only adds to the new unbonded list if it hasn't
+			// previously unbonded with types.BTCDelegationStatus_UNBONDED
+			k.processPowerDistUpdateEventUnbond(ctx, state, btcDel)
+		}
+	}
+}
+
+func (k Keeper) processFPEventImmediate(ctx sdk.Context, state *ftypes.ProcessingState, event types.EventPowerDistUpdate) {
+	switch typedEvent := event.Ev.(type) {
+	case *types.EventPowerDistUpdate_SlashedFp:
```

Comment:
maybe lets double check that there is no operations leading to following sequences:
1. Slashed event
2. Jailed event

or

1. Slashed event
2. Unjail event

as this would mean we will not correctly process `Slashing` event  later down the line i.e we won't do
```go
		case ftypes.FinalityProviderState_SLASHED:
			// if this finality provider is slashed, continue to avoid
			// assigning delegation to it
			fp.IsSlashed = true
			continue
```
which means guy will remain part of voting `VotingPowerDistCache`

Reply:
In the code:
- `[JailFinalityProvider](https://github.com/babylonlabs-io/babylon/blob/91403d33146fc5b1a0b0a0e68c421339fa5fb12f/x/btcstaking/keeper/finality_providers.go#L250-L251)` checks if FP is slashed. So **not** possible to have the sequence SLASHED -> JAILED
- `SlashFinalityProvider` does not check if is the FP is jailed. 
So could happen that a jailed FP is slashed.
- `[UnjailFinalityProvider](https://github.com/babylonlabs-io/babylon/blob/91403d33146fc5b1a0b0a0e68c421339fa5fb12f/x/btcstaking/keeper/finality_providers.go#L285-L286)` does not check if FP is slashed, so could be unjailed
Thus, the sequence SLASHED -> UNJAILED seems to be possible

Reply:
considering SLASHED is the terminal state for FPs, I'm aligned with @RafilxTenfen 's idea of deferring the slashed events (as we do with the EXPIRED delegations)

Reply:
> unless we add a check in the UnjailFinalityProvider to avoid unjailing slashed FPs
that also makes sense to me

Reply:
Hmm I am slightly leaning to:
> considering this, and that SLASHED is the terminal state for FPs, I'm aligned with @RafilxTenfen 's idea of deferring the slashed events (as we do with the EXPIRED delegations)
mostly because this better matches semantics of slashing i.e if slashing happens it should not be possible to do anything to change this state. Like the switch that you can flick one time.

So if we were to introduce yet another state for fp in the future, slashing should still be the one overriding this state.

Reply:
updated with these changes